Rework DayNight - reduce object creation

rhuanjl opened the pull request:
This isn't as pretty but it is functionally identical, faster and doesn't generate and discard objects - therefore reduces GC pressure.

Significant changes:
1. Do the colours with a model and a shader (gets rid of Prim use and no Colour.mix object creation)

2. Use Date.now() instead of new date - gives milliseconds since 1st Jan 1970 - considering the intent is time in an individual day effect is the same and doesn't instantiate an object

fatcerberus commented on src/dayNightSystem/dayNightClock.mjs:
```diff
@@ -6,69 +6,101 @@
 
 		console.log("initializing day/night clock", `time: ${this.now()}`);
-		this.currentMask = Color.Transparent;
+		let white = Color.White;
```
This line could be removed entirely: White is the default color for vertices if no color is specified, exactly so that you can make colorizable "silhouettes" like you're doing here.

rhuanjl replied:
removed

fatcerberus commented on src/dayNightSystem/dayNightClock.mjs:
```diff
@@ -6,69 +6,101 @@
-      NightMask = new Color(0.0, 0.0, 0.125, 0.5625);
+const DayMask = [0.0, 0.0, 0.0, 0.0],
+      TwilightMask = [0.5, 0.125, 0.0625, 0.625],
+      NightMask = [0.0, 0.0, 0.125, 0.5625];
```
Nit: Spaces between `[]` and values

rhuanjl replied:
revised to your style

fatcerberus commented on src/dayNightSystem/dayNightClock.mjs:
```diff
@@ -6,69 +6,101 @@
+			fragmentFile: '#/shaders/image.frag.glsl',
+			vertexFile  : '#/shaders/image.vert.glsl',
+		});
+		this.model.shader = this.shader;
```
Maybe we should just create the shader before the model and do `new Model([ rectangle ], shader);` so we don't need to set it afterwards.

rhuanjl replied:
done

fatcerberus commented on src/dayNightSystem/dayNightClock.mjs:
```diff
@@ -6,69 +6,101 @@
-		let second = currentTime % 60;
-		return new InGameTime(hour, minute, second);
+		const time = Date.now();
+		let currentTime = Math.floor(time / 100) % 86400;
```
I'm not sure if this is 100% equivalent to what I had before?  To clarify, how this should work is: A Lucidan "week" is 10 in-game days long and always begins at the start of a waking-world day, which is to say: midnight in realtime = midnight of day 1 in-game.  As your version is just "seconds since 1/1/1970, modulo 86400" I think that might put the cycle out of sync?

fatcerberus replied:
I verified against master, and I was right--the sync is off.  As of this message the version in master reports ~21:00 (night) while yours reports ~13:00 (day)

fatcerberus commented on src/dayNightSystem/dayNightClock.mjs:
```diff
@@ -6,69 +6,101 @@
+		this.currentMask[0] = mask1[0] * proportion1 + mask2[0] * proportion2;
+		this.currentMask[1] = mask1[1] * proportion1 + mask2[1] * proportion2;
+		this.currentMask[2] = mask1[2] * proportion1 + mask2[2] * proportion2;
+		this.currentMask[3] = mask1[3] * proportion1 + mask2[3] * proportion2;
```
I wonder if we could do this in a shader somehow...

rhuanjl replied:
I can't think of an efficient way to put this in a shader - any method I can think of would involve multiple uniforms and be  more costly overal.

It could perhaps be re-factored as some kind of more widely available helper function or method though rather than being just here.

fatcerberus commented:
I just remembered that there's a bug in master where the first frame after the DayNightClock starts running will not apply the correct filter as the mask gets set to `Color.Transparent` in the constructor:
https://github.com/fatcerberus/spectacles-i/blob/master/src/dayNightSystem/dayNightClock.mjs#L25

The event loop does renders before updates, so `on_update()` doesn't get a chance to set it properly for that first frame.  Could you fix that here too?  In a separate commit, preferably.

rhuanjl commented:
I've fixed your points.

I've added an update call too the constructor to dodge the first frame issue.

I've also put in an offset so it starts synced to the players timezone. Note if maintaining timezone sync is vital may have to put in a periodic check for daylight saving time changes and/or the user having changed their computer's timezone.
